crates/red_knot_workspace/src/workspace.rs

```diff
@@ -186,35 +183,45 @@ impl Workspace {
     }
 
     /// Checks all open files in the workspace and its dependencies.
-    pub fn check(self, db: &RootDatabase) -> Vec<Box<dyn Diagnostic>> {
-        let workspace_span = tracing::debug_span!("check_workspace");
-        let _span = workspace_span.enter();
-
-        tracing::debug!("Checking workspace");
-        let files = WorkspaceFiles::new(db, self);
-        let result = Arc::new(std::sync::Mutex::new(Vec::new()));
-        let inner_result = Arc::clone(&result);
-
-        let db = db.snapshot();
-        let workspace_span = workspace_span.clone();
-
-        rayon::scope(move |scope| {
-            for file in &files {
-                let result = inner_result.clone();
-                let db = db.snapshot();
-                let workspace_span = workspace_span.clone();
-
-                scope.spawn(move |_| {
-                    let check_file_span = tracing::debug_span!(parent: &workspace_span, "check_file", file=%file.path(&db));
-                    let _entered = check_file_span.entered();
-
-                    let file_diagnostics = check_file(&db, file);
-                    result.lock().unwrap().extend(file_diagnostics);
-                });
-            }
+    pub fn check(self, db: &RootDatabase) -> Vec<CompileDiagnostic> {
+        check_workspace_par(db, self);
+
+        let mut diagnostics: Vec<CompileDiagnostic> =
+            check_workspace_sync::accumulated::<CompileDiagnostic>(db, self);
+
+        diagnostics.sort_unstable_by(|a, b| {
+            let a_file = a.file();
+            let b_file = b.file();
+
+            a_file.cmp(&b_file).then_with(|| {
+                a_file
+                    .path(db)
+                    .as_str()
+                    .cmp(b_file.path(db).as_str())
+                    .then_with(|| {
+                        a.range()
+                            .unwrap_or_default()
+                            .start()
+                            .cmp(&b.range().unwrap_or_default().start())
+                    })
+            })
```

Comment:
A few questions:

- Do we really want an unstable sort here? Two different diagnostics could compare equal according to the ordering defined here if they refer to the same range, or even just the same range start.
- Why do we order by file and then by the path of the file? Can multiple files refer to the same path?
- If performance is not a major concern here, we could potentially do something like
  ```rs
  diagnostics.sort_unstable_by_key(|diagnostic| {
      let file = diagnostic.file();
      let path = file.path(db).as_str();
      let range_start = diagnostic.range().unwrap_or_default().start();
      (file, path, range_start)
  });
  ```

Reply:
> Why do we order by file and then by the path of the file? Can multiple files refer to the same path?

Oh, that's just wrong. The idea was to short-cut if the files are identical. 

> Do we really want an unstable sort here? Two different diagnostics could compare equal according to the ordering defined here if they refer to the same range, or even just the same range start.

I don't think it's important for us to preserve the original order if there are multiple diagnostics at the same line, for as long as the order is deterministic (Running Red Knot multiple times should give you the same result). We should probably include the rule code as well as disambiguator

Reply:
> for as long as the order is deterministic

Exactly, that was my concern.

> We should probably include the rule code as well as disambiguator

:+1: